tests/functional/test_discovery.py

```diff
@@ -0,0 +1,371 @@
+            stubber.add_response(status=200, body=b'{}')
+            client.test_discovery_optional(Foo="bar")
+            self.assert_discovery_skipped(
+                stubber,
+                b"test-discovery-endpoint.TestDiscoveryOptional"
+            )
+            self.assert_endpoint_used(stubber.requests[0].url, endpoint)
+
+    def test_endpoint_discovery_optional_disabled_with_custom_endpoint(self):
+        endpoint = "https://custom.domain/"
+        config = Config(endpoint_discovery_enabled=False)
+        client, http_stubber = self.create_client(
+            service_name="test-discovery-endpoint",
+            config=config,
+            endpoint_url=endpoint
+        )
+        with http_stubber as stubber:
+            stubber.add_response(status=200, body=b'{}')
+            client.test_discovery_optional(Foo="bar")
+            self.assert_discovery_skipped(
+                stubber,
+                b"test-discovery-endpoint.TestDiscoveryOptional",
+            )
+            self.assert_endpoint_used(stubber.requests[0].url, endpoint)
+
+    def test_endpoint_discovery_default_optional_endpoint(self):
+        client, http_stubber = self.create_client(service_name="test-discovery-endpoint")
+        with http_stubber as stubber:
+            stubber.add_response(status=200, body=b'{}')
+            client.test_discovery_optional(Foo="bar")
+            self.assertEqual(len(stubber.requests), 1)
+
+    def test_endpoint_discovery_enabled_optional_endpoint(self):
+        discovered_endpoint = 'https://discovered.domain'
+        config = Config(endpoint_discovery_enabled=True)
+        client, http_stubber = self.create_client(
+            service_name="test-discovery-endpoint", config=config
+        )
+        with http_stubber as stubber:
+            self.add_describe_endpoints_response(stubber, discovered_endpoint)
+            client.test_discovery_optional(Foo="bar")
+            self.assert_endpoint_discovery_used(stubber, discovered_endpoint)
+
+    def test_endpoint_discovery_manual_auto_on_required_endpoint(self):
+        discovered_endpoint = 'https://discovered.domain'
+        config = Config(endpoint_discovery_enabled=" aUto  \n")
+        client, http_stubber = self.create_client(
+            service_name="test-discovery-endpoint", config=config
+        )
+        with http_stubber as stubber:
+            self.add_describe_endpoints_response(stubber, discovered_endpoint)
+            client.test_discovery_required(Foo="bar")
+            self.assert_endpoint_discovery_used(stubber, discovered_endpoint)
+
+    def test_endpoint_discovery_enabled_with_random_string(self):
+        config = Config(endpoint_discovery_enabled="bad value")
+        client, http_stubber = self.create_client(
+            service_name="test-discovery-endpoint", config=config
+        )
+        http_stubber.add_response(status=200, body=b'{}')
+        with self.assertRaises(EndpointDiscoveryRequired):
```

Comment:
I feel like this should instead be a validation error for a bad configuration value instead.

Reply:
I agree this could be more direct, this was hold over from the previous behavior for bad values. Let me see if I can add some better handling.